File: 51/Solution.java
import java.util.ArrayList;

import com.sun.xml.internal.bind.v2.schemagen.xmlschema.List;

class Solution {
    private int n;
    List<List<String>> answers;
    private int[] hasQueue;                                  // input the index of rows to get the queen's column~
    private String rowChar;
    public List<List<String>> solveNQueens(int n) {
        this.n = n;
        hasQueue = new int[n];
        answers = new ArrayList<>();
        rowChar = new String("");
        for (int i = 0; i < n; i++) {
            rowChar+='.';
            hasQueue[i] = -n-2;
        }
        FindQueen(0);
        return answers;

    }
    public boolean isValid(int row,int col)
    {
        for (int i = 0; i < n; i++) {
            if(hasQueue[i]==col || Math.abs(row-i)==Math.abs(col-hasQueue[i]))
            {
                return false;
            }
        
        }
        return true;
    }
    public void FindQueen(int row)
    {
        if(row == n)                                        // 得到新解返回
        {
            List<String> oneAnswer = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                StringBuilder sb = new StringBuilder(rowChar);
                sb.setCharAt(hasQueue[i], 'Q');
                oneAnswer.add(sb.toString());
            }
            answers.add(oneAnswer);
            return ;
        }
        for (int i = 0; i < n; i++) {
            if (isValid(row,i)) {
                hasQueue[row]=i;
                FindQueen(row+1);
                hasQueue[row]=(-n-2);
            }
        }

    }
}

File: 51/Solution.py
class Solution:
    # n=-1                # n皇后
    # answers = []     # 存放所有answer
    # hasQueue=[]      # 存放第i行对应皇后的列索引位，  (i,hasQueue[i]) 是皇后的坐标位置
    def solveNQueens(self, n):
        self.answers=[]
        self.n=n
        self.hasQueue = [-n-1]*n
        self.findQueen(0)
        return self.answers
    

    def isValid(self,row,col):                       # 判断（row,col）位置是否可以放皇后
        for i in range(self.n):
            if col==self.hasQueue[i] or abs(row-i) == abs(col-self.hasQueue[i]):
                return False
        return True
    
    def findQueen(self,row):
        if row==self.n:
            oneAnswer = []
            for Q in self.hasQueue:
                line = ['.']*self.n
                line[Q] = 'Q'
                line = "".join(line)
                oneAnswer.append(line)
            self.answers.append(oneAnswer)
            return 
        for i in range(self.n):
            if self.isValid(row,i):
                self.hasQueue[row] = i
                self.findQueen(row+1)
                self.hasQueue[row] = -self.n-1    # 回溯
        
